File: 128.longest-consecutive-sequence.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=128 lang=java
 *
 * [128] Longest Consecutive Sequence
 */

// @lc code=start
class Solution {

class UnionFind{
int[] size;
int[] id;

    public UnionFind(int size){

        this.id = new int[size];
        this.size = new int[size];
        for(int i = 0; i < size; i++){
            this.id[i] = i;
            this.size[i] = 1;
        }
    }

    public  int find(int x){

        int root = x;
        while(root != id[root]){
            root = id[root];
        }
        //compress path
        while(x != root){
            int temp = id[x];
            id[x] = root;
            x = temp;
        }
        return root;

    }
    public void union(int x, int y){
        int rootX = find(x);
        int rootY = find(y);
        if(rootX == rootY){
            return;
        }
        if(size[rootX] > size[rootY]){
            id[rootY] = rootX;
            size[rootX] += size[rootY];
        }else{
            id[rootX] = rootY;
            size[rootY] += size[rootX];
        }
    }
    public int findLargestComponent(){
        int max = 0;
        for(int i = 0; i < size.length; i++){
            max = Math.max(max, size[i]);
        }
        return max;
    }


}

    public int longestConsecutive(int[] nums) {
        UnionFind uf = new UnionFind(nums.length);
        Map<Integer,Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if(map.containsKey(nums[i])){
                continue;
            }
            if(map.containsKey(nums[i] - 1)){
                uf.union(i, map.get(nums[i] - 1));
            }
            if(map.containsKey(nums[i] + 1)){
                uf.union(i, map.get(nums[i] + 1));
            }
            map.put(nums[i], i);
        }
        return uf.findLargestComponent();
        
    }
}
// @lc code=end

File: 128.longest-consecutive-sequence.py
#
# @lc app=leetcode id=128 lang=python3
#
# [128] Longest Consecutive Sequence
#

# @lc code=start
from typing import List


class Solution:

        class UnionFind:
            
            def __init__(self, size):
                self.count = size
                self.size = [1 for _ in range(size)]
                self.id = list(range(size))
            
            def find(self,p)-> int:
                root = p
                while self.id[root] != root:
                    root = self.id[root]
                
                #path compression
                while root != p:
                    temp = self.id[p]
                    self.id[p] = root
                    p = temp
                    
                return root
            
            def union(self, p, q) -> bool:
                rootP = self.find(p)
                rootQ = self.find(q)
                if self.isConnected(p, q):
                    return False
                if self.size[rootP] > self.size[rootQ]:
                    self.id[rootQ] = rootP
                    self.size[rootP] += self.size[rootQ]
                else:
                    self.id[rootP] = rootQ
                    self.size[rootQ] += self.size[rootP]
                self.count -= 1
                return True
            
            def getSize(self,p)-> int:
                return self.size[self.find(p)]
            
            def isConnected(self,p,q)-> bool:
                return self.find(p) == self.find(q)
            
            def getId(self,p)-> int:
                return self.id[p]
            
            def findLargestComponent(self) -> int:
                return self.find(self.size.index(max(self.size)))
            
            def getSizeOfLargestComponent(self) -> int:
                return max(self.size)
            
            def numberOfComponents(self) -> int:
                return self.count
            
            
        def longestConsecutive(self, nums: List[int]) -> int:
            
            
        
# @lc code=end

